File: packages/openmemory-py/src/openmemory/utils/chunking.py
import math
import re
from typing import List, Dict, TypedDict, Optional
try:
    import tiktoken
    _HAS_TIKTOKEN = True
except ImportError:
    _HAS_TIKTOKEN = False

class Chunk(TypedDict):
    text: str
    start: int
    end: int
    tokens: int

def get_tokenizer():
    if _HAS_TIKTOKEN:
        try:
            return tiktoken.get_encoding("cl100k_base")
        except (ValueError, RuntimeError):
            return None
    return None

_tokenizer = get_tokenizer()

def est_tokens(t: str) -> int:
    if _tokenizer:
        # Use disallowed_special=() to treat special tokens as normal text
        return len(_tokenizer.encode(t, disallowed_special=()))
    # Fallback to character-based estimate if tiktoken is not available
    return math.ceil(len(t) / 4)

def chunk_text(txt: str, tgt: int = 768, ovr: float = 0.1) -> List[Chunk]:
    tot = est_tokens(txt)
    if tot <= tgt:
        return [{"text": txt, "start": 0, "end": len(txt), "tokens": tot}]

    tch = tgt * 4
    och = math.floor(tch * ovr)
    paras = re.split(r"\n\n+", txt)

    chks: List[Chunk] = []
    cur = ""
    cs = 0
    cur_tokens = 0

    for p in paras:
        sents = re.split(r"(?<=[.!?])\s+", p)
        for s in sents:
            s_tokens = est_tokens(s)
            pot_tokens = cur_tokens + (1 if cur else 0) + s_tokens

            if pot_tokens > tgt and len(cur) > 0:
                chks.append({
                    "text": cur,
                    "start": cs,
                    "end": cs + len(cur),
                    "tokens": cur_tokens
                })
                ovt = cur[-och:] if och < len(cur) else cur
                ovt_tokens = est_tokens(ovt)
                cur = ovt + " " + s
                cs = cs + len(cur) - len(ovt) - 1
                cur_tokens = ovt_tokens + 1 + s_tokens
            else:
                cur = cur + (" " if cur else "") + s
                cur_tokens = pot_tokens

    if len(cur) > 0:
        chks.append({
            "text": cur,
            "start": cs,
            "end": cs + len(cur),
            "tokens": cur_tokens
        })
    return chks

def agg_vec(vecs: List[List[float]]) -> List[float]:
    n = len(vecs)
    if not n: raise ValueError("no vecs")
    if n == 1: return vecs[0].copy()

    d = len(vecs[0])
    r = [0.0] * d
    for v in vecs:
        for i in range(d):
            r[i] += v[i]

    rc = 1.0 / n
    for i in range(d):
        r[i] *= rc
    return r

def join_chunks(cks: List[Chunk]) -> str:
    return " ".join(c["text"] for c in cks) if cks else ""

File: packages/openmemory-js/src/utils/chunking.ts
import { getEncoding, Tiktoken } from "js-tiktoken";

export type chunk = {
    text: string;
    start: number;
    end: number;
    tokens: number;
};

let encoder: Tiktoken | null = null;

function get_encoder(): Tiktoken | null {
    if (encoder) return encoder;
    try {
        encoder = getEncoding("cl100k_base");
        return encoder;
    } catch (e) {
        console.warn("[CHUNKING] Failed to initialize tiktoken encoder, falling back to character estimate:", e);
        return null;
    }
}

const est = (t: string) => {
    const enc = get_encoder();
    if (enc) {
        // Use disallowedSpecial: [] to treat special tokens as normal text
        return enc.encode(t, "all", []).length;
    }
    return Math.ceil(t.length / 4);
};

/**
 * Splits a paragraph into chunks based on token limit and overlap.
 */
function process_paragraph_sentences(
    p: string,
    chks: chunk[],
    tgt: number,
    och: number,
    state: { cur: string; cs: number; cur_tokens: number }
) {
    const sents = p.split(/(?<=[.!?])\s+/);
    for (const s of sents) {
        const s_tokens = est(s);
        const pot_tokens = state.cur_tokens + (state.cur ? 1 : 0) + s_tokens;

        if (pot_tokens > tgt && state.cur.length > 0) {
            chks.push({
                text: state.cur,
                start: state.cs,
                end: state.cs + state.cur.length,
                tokens: state.cur_tokens,
            });
            const ovt = state.cur.slice(-och);
            const ovt_tokens = est(ovt);
            state.cur = ovt + " " + s;
            state.cs = state.cs + state.cur.length - ovt.length - 1;
            state.cur_tokens = ovt_tokens + 1 + s_tokens;
        } else {
            state.cur = state.cur + (state.cur ? " " : "") + s;
            state.cur_tokens = pot_tokens;
        }
    }
}

export const chunk_text = (txt: string, tgt = 768, ovr = 0.1): chunk[] => {
    const tot = est(txt);
    if (tot <= tgt)
        return [{ text: txt, start: 0, end: txt.length, tokens: tot }];

    const tch = tgt * 4,
        och = Math.floor(tch * ovr);
    const paras = txt.split(/\n\n+/);

    const chks: chunk[] = [];
    const state = { cur: "", cs: 0, cur_tokens: 0 };

    for (const p of paras) {
        process_paragraph_sentences(p, chks, tgt, och, state);
    }

    if (state.cur.length > 0)
        chks.push({
            text: state.cur,
            start: state.cs,
            end: state.cs + state.cur.length,
            tokens: state.cur_tokens,
        });
    return chks;
};

export const agg_vec = (vecs: number[][]): number[] => {
    const n = vecs.length;
    if (!n) throw new Error("no vecs");
    if (n === 1) return vecs[0].slice();

    const d = vecs[0].length,
        r = new Array(d).fill(0);
    for (const v of vecs) for (let i = 0; i < d; i++) r[i] += v[i];
    const rc = 1 / n;
    for (let i = 0; i < d; i++) r[i] *= rc;
    return r;
};

export const join_chunks = (cks: chunk[]) =>
    cks.length ? cks.map((c) => c.text).join(" ") : "";
